node/src/main/scala/org/bitcoins/node/PeerManager.scala

```diff
@@ -1,199 +1,330 @@
+        throw new IllegalArgumentException(
+          s"Unsupported address type of size $unknownSize bytes")
     }
+    PeerDAO()
+      .upsertPeer(ByteVector(addrBytes),
+                  peer.socket.getPort,
+                  networkByte,
+                  serviceIdentifier)
   }
 
-  /** Returns peers randomly taken from config, database, hardcoded peers, dns seeds in that order */
-  def getPeers: Future[Vector[Peer]] = {
-    //currently this would only give the first peer from config
-    val peersFromConfig = getPeersFromConfig
-    val peersFromDbF = getPeersFromDb
-    val peersFromResources = getPeersFromResources
-    val maxConnectedPeers = nodeAppConfig.maxConnectedPeers
-
-    val allF = for {
-      peersFromDb <- peersFromDbF
-    } yield {
-      val shuffledPeers = (Random.shuffle(peersFromConfig) ++ Random.shuffle(
-        peersFromDb) ++ Random.shuffle(peersFromResources)).distinct
-
-      //getting peers from dns seeds takes a noticeable 5-8 sec so treating this separately
-      if (maxConnectedPeers > shuffledPeers.size) {
-        shuffledPeers.take(maxConnectedPeers) ++ getPeersFromDnsSeeds.take(
-          maxConnectedPeers - shuffledPeers.size)
-      } else {
-        shuffledPeers.take(maxConnectedPeers)
+  def awaitPeerWithService(
+      f: ServiceIdentifier => Boolean,
+      timeout: Duration): Future[Unit] = {
+    logger.debug(s"Waiting for peer connection. ${_peerData.keys}")
+    val ret = AsyncUtil
+      .retryUntilSatisfied({
+                             _peerData.exists(x => f(x._2.serviceIdentifier))
+                           },
+                           interval = 1.seconds,
+                           maxTries = timeout.getSeconds.toInt)
+      .recover {
+        case _: AsyncUtil.RpcRetryException =>
+          throw new RuntimeException("No supported peers found!")
+        case unknown: Throwable => throw unknown
       }
-    }
-    allF
+
+    ret
   }
 
-  def addPeer(peer: Peer): Unit = {
-    if (!_peerData.contains(peer))
-      _peerData.put(peer, PeerData(peer, node))
-    else logger.debug(s"Peer $peer already added.")
-    ()
+  def replacePeer(replacePeer: Peer, withPeer: Peer): Future[Unit] = {
+    logger.debug(s"Replacing $replacePeer with $withPeer")
+    assert(!peerData(replacePeer).serviceIdentifier.nodeCompactFilters,
+           s"$replacePeer has cf")
+    removePeer(replacePeer)
+    addPeer(withPeer)
   }
 
-  def removePeer(peer: Peer): Future[Unit] = {
-    if (_peerData.contains(peer)) {
-      val connF = peerData(peer).peerMessageSender.isConnected()
-      val disconnectF = connF.map { conn =>
-        if (conn) peerData(peer).peerMessageSender.disconnect()
-        else Future.unit
+  def removePeer(peer: Peer): Unit = {
+    logger.debug(s"Removing persistent peer $peer")
+    val client = peerData(peer).client
+    _peerData.remove(peer)
+    //so we need to remove if from the map for connected peers so no more request could be sent to it but we before
```

Comment:
When we don't want a peer as one of our persistent ones, we remove it from the `peerData` map, but still keep it around to check if the client actor for it actually stopped.